parlai/core/metrics.py

```diff
@@ -78,32 +89,33 @@ def as_number(self, obj: TScalar) -> Union[int, float]:
         assert isinstance(obj_as_number, int) or isinstance(obj_as_number, float)
         return obj_as_number
 
-    def as_float(self, obj: Union[int, float, torch.Tensor]) -> float:
-        obj_as_number = self.as_number(obj)
-        return float(obj_as_number)
+    def as_float(self, obj: TScalar) -> float:
+        return float(self.as_number(obj))
 
-    def as_int(self, obj: Union[int, float, torch.Tensor]) -> int:
-        obj_as_number = self.as_number(obj)
-        if isinstance(obj_as_number, float):
-            assert obj_as_number.is_integer()
-            return int(obj_as_number)
-        else:
-            assert isinstance(obj_as_number, int)
-            return obj_as_number
+    def as_int(self, obj: TScalar) -> int:
+        return int(self.as_number(obj))
 
 
 class SumMetric(Metric):
     """
     Class that keeps a running sum of some metric.
     """
 
-    def __init__(self, sum_: TScalar):
-        self._sum = self.as_float(sum_)
+    __slots__ = ('_sum',)
+
+    def __init__(self, sum_: TScalar = 0):
+        if isinstance(sum_, torch.Tensor):
+            self._sum = sum_.item()
+        else:
+            assert isinstance(sum_, (int, float))
+            self._sum = sum_
 
-    def __add__(self, other: 'SumMetric') -> 'SumMetric':
+    def __add__(self, other: Optional['SumMetric']) -> 'SumMetric':
         # NOTE: hinting can be cleaned up with "from __future__ import annotations" when
         # we drop Python 3.6
-        full_sum: float = self._sum + other._sum
+        if other is None:
```

Comment:
Maybe good to add a unit test for other = None? Not sure how thorough we need to be :)

Reply:
I don't think so. I'm not interested in that case in principal. It's just a convenience factor for the user of the class, and if that became no longer necessary, I wouldn't feel a clear need for that case to be covered.